server/src/main/java/com/microsoft/java/bs/core/internal/services/BuildTargetService.java

```diff
@@ -469,6 +477,86 @@ public ScalacOptionsResult getBuildTargetScalacOptions(ScalacOptionsParams param
+      StatusCode statusCode;
+      if (TestParamsDataKind.SCALA_TEST.equals(params.getDataKind())) {
+        // ScalaTestParams is for a list of classes only
+        ScalaTestParams testParams = JsonUtils.toModel(params.getData(), ScalaTestParams.class);
+        Map<BuildTargetIdentifier, Map<String, Set<String>>> testClasses = new HashMap<>();
+        for (ScalaTestClassesItem testClassesItem : testParams.getTestClasses()) {
+          Map<String, Set<String>> classesMethods = new HashMap<>();
+          for (String classNames : testClassesItem.getClasses()) {
+            classesMethods.put(classNames, null);
+          }
+          testClasses.put(testClassesItem.getTarget(), classesMethods);
+        }
+        statusCode = connector.runTests(entry.getKey(), testClasses, testParams.getJvmOptions(),
+            params.getArguments(), null, client, params.getOriginId(),
+            compileProgressReporter);
+      } else if ("scala-test-suites-selection".equals(params.getDataKind())) {
+        // ScalaTestSuites is for a list of classes + methods
+        // Since it doesn't supply the specific BuildTarget we require a single
+        // build target in the params and reject any request that doesn't match this
+        if (params.getTargets().size() != 1) {
+          LOGGER.warning("Test params with Test Data Kind " + params.getDataKind()
+              + " must contain only 1 build target");
+          statusCode = StatusCode.ERROR;
+        } else {
+          ScalaTestSuites testSuites = JsonUtils.toModel(params.getData(), ScalaTestSuites.class);
+          Map<String, String> envVars = null;
+          boolean argsValid = true;
+          if (testSuites.getEnvironmentVariables() != null) {
+            // arg is of the form KEY=VALUE
+            List<String[]> splitArgs = testSuites.getEnvironmentVariables()
+                .stream()
+                .map(arg -> arg.split("="))
+                .collect(Collectors.toList());
+            argsValid = splitArgs.stream().allMatch(arg -> arg.length == 2);
+            if (argsValid) {
+              envVars = splitArgs.stream().collect(Collectors.toMap(arg -> arg[0], arg -> arg[1]));
+            }
+          }
+          if (!argsValid) {
+            LOGGER.warning("Test params arguments must each be in the form KEY=VALUE. "
+                + testSuites.getEnvironmentVariables());
+            statusCode = StatusCode.ERROR;
+          } else {
+            Map<BuildTargetIdentifier, Map<String, Set<String>>> testClasses = new HashMap<>();
+            for (ScalaTestSuiteSelection testSuiteSelection : testSuites.getSuites()) {
+              Map<String, Set<String>> classesMethods = new HashMap<>();
+              Set<String> methods = new HashSet<>(testSuiteSelection.getTests());
+              classesMethods.put(testSuiteSelection.getClassName(), methods);
+              testClasses.put(params.getTargets().get(0), classesMethods);
```

Comment:
The `classesMethods` is always re-initialized at line 538, makes the testClasses always only contain one test target.

Reply:
Yes - it's tricky.  The `TestParams` passes a list of build target ids and a list of test classes.  I've got no way of associating these, i.e. working out which test class applies to which build target.  But luckily Gradle doesn't care - you just supply the classes to test regardless of project.  But we care because `TestReportReporter` has to pass back a build target id.  So I've used the first build target id in all cases.  I'll have a look to see if there's any way to work this out correctly from the Gradle test events but I doubt it.  I assume you don't actually care about build target ids at the BSP client end - you just use the test name data?

Reply:
> I assume you don't actually care about build target ids at the BSP client end - you just use the test name data?

Yes, my client implementation makes sure that only one build target will be included in the request.

Reply:
Hi @Arthurm1, what about this? Assume the client will only send one target now?